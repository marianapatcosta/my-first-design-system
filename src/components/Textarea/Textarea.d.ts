import React, { ChangeEvent, CSSProperties } from 'react';
export interface TextareaProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional css class
     */
    className?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
    /**
     * Textarea label
     */
    label?: string;
    /**
     * Textarea name
     */
    name?: string;
    /**
     * Textarea value
     */
    value: string;
    /**
     * Textarea's displaying text when no option is selected
     */
    placeholder?: string;
    /**
     * Is Textarea disabled?
     */
    disabled?: boolean;
    /**
     * what is rows number allowed for the Textarea field?
     */
    rows?: number;
    /**
     * text to display in case of error
     */
    errorMessage?: string;
    /**
     * Change handler
     */
    onChange: (event: ChangeEvent<HTMLTextAreaElement>) => void;
    /**
     * Blur handler
     */
    onBlur?: (event: ChangeEvent<HTMLTextAreaElement>) => void;
}
export declare const Textarea: React.FC<TextareaProps>;
