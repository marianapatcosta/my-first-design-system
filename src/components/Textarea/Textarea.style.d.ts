export declare const StyledTextareaWrapper: import("styled-components").StyledComponent<"div", any, {}, never>;
export declare const StyledTextarea: import("styled-components").StyledComponent<"textarea", any, {
    errorMessage?: string | undefined;
    disabled?: boolean | undefined;
}, never>;
export declare const StyledLabel: import("styled-components").StyledComponent<"label", any, {
    disabled?: boolean | undefined;
}, never>;
export declare const StyledTextareaError: import("styled-components").StyledComponent<"p", any, {}, never>;
