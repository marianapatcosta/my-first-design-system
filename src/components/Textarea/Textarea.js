var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { StyledTextareaWrapper, StyledTextarea, StyledLabel, StyledTextareaError, } from './Textarea.style';
export var Textarea = function (_a) {
    var id = _a.id, label = _a.label, disabled = _a.disabled, errorMessage = _a.errorMessage, placeholder = _a.placeholder, props = __rest(_a, ["id", "label", "disabled", "errorMessage", "placeholder"]);
    return (_jsxs(StyledTextareaWrapper, { children: [label && (_jsx(StyledLabel, __assign({ htmlFor: id, disabled: disabled }, { children: label }), void 0)),
            _jsx(StyledTextarea, __assign({ id: id, "data-testid": 'textarea-field', "aria-label": !!label ? label : placeholder, placeholder: placeholder, errorMessage: errorMessage, disabled: disabled }, props), void 0),
            !!errorMessage && (_jsx(StyledTextareaError, { children: errorMessage }, void 0))] }, void 0));
};
Textarea.defaultProps = {
    id: '',
    className: '',
    style: {},
    name: '',
    rows: 5,
    disabled: false,
    value: '',
    placeholder: '',
    errorMessage: '',
};
