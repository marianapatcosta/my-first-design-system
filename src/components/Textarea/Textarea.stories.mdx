import { Meta, Story, ArgsTable, Canvas, Preview } from '@storybook/addon-docs/blocks';
import { Textarea } from "./Textarea";

<Meta title="Design System/Textarea" component={Textarea} />

# Textarea
<br/><br/> 

### Displays a customized Textarea field, to insert large amount of data

Use textarea when the data inserted by the user may require multi-line written information

<Preview withToolbar>
  <Story name="Standard">
    <Textarea value="hello" />
    />
  </Story>
</Preview>

<ArgsTable of={Textarea} />

### Textarea with Label:

  <Story name="WithIcon">
    <Textarea value="hello" label="Input" />
  </Story>

### Textarea with Error:
  <Story name="WithError">
    <Textarea value="hello" errorMessage="This is an error!" />
  </Story>

### Playground

Experiment with Textarea Component with the Controls addon in the Canvas tab.

export const Template = (args) => <Textarea {...args} />

<Canvas>
  <Story name="Controls" args={{
  value: 'hello'
  }}>
    {Template.bind({})}
  </Story>
</Canvas> 