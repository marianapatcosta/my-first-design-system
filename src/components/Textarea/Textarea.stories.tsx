import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { Textarea, TextareaProps } from './Textarea';

export default {
  title: 'Textarea',
  component: Textarea,
  args: { value: 'hello' },
} as Meta;

const Template: Story<TextareaProps> = (args) => <Textarea {...args} />;

export const Default = Template.bind({});
Default.args = {};

export const WithLabel = Template.bind({});
WithLabel.args = { label: 'Textarea' };

export const WithError = Template.bind({});
WithError.args = { errorMessage: 'This is an error!' };
