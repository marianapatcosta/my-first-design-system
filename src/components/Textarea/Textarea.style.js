var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled from 'styled-components';
import { baseStyles, disabledStyles, colors, typography } from '../../styles';
export var StyledTextareaWrapper = styled.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n  margin: 1rem 0;\n  position: relative;\n  width: 100%;\n"], ["\n  ", "\n  margin: 1rem 0;\n  position: relative;\n  width: 100%;\n"])), baseStyles);
export var StyledTextarea = styled.textarea(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  display: block;\n  color: ", ";\n  font-family: ", ";\n  font-size: 95%;\n  width: 100%;\n  width: 100%;\n  background-color: ", ";\n  border: 0.0625rem solid ", ";\n  box-shadow: 0 0.063rem 0.25rem ", ";\n  box-sizing: border-box;\n  padding: 0.3rem 0.625rem;\n  position: relative;\n  resize: none;\n\n  :-webkit-autofill,\n  :-webkit-autofill:hover,\n  :-webkit-autofill:focus,\n  :-webkit-autofill:active {\n    -webkit-text-fill-color: ", ";\n    -webkit-box-shadow: 0 0 0 2rem ", " inset !important;\n  }\n\n  ", "\n\n  ", "\n\n  :focus-visible {\n    outline: 0.125rem solid ", ";\n  }\n\n  :focus:not(:focus-visible) {\n    outline: none;\n  }\n"], ["\n  display: block;\n  color: ", ";\n  font-family: ", ";\n  font-size: 95%;\n  width: 100%;\n  width: 100%;\n  background-color: ", ";\n  border: 0.0625rem solid ", ";\n  box-shadow: 0 0.063rem 0.25rem ", ";\n  box-sizing: border-box;\n  padding: 0.3rem 0.625rem;\n  position: relative;\n  resize: none;\n\n  :-webkit-autofill,\n  :-webkit-autofill:hover,\n  :-webkit-autofill:focus,\n  :-webkit-autofill:active {\n    -webkit-text-fill-color: ", ";\n    -webkit-box-shadow: 0 0 0 2rem ", " inset !important;\n  }\n\n  ", "\n\n  ",
    "\n\n  :focus-visible {\n    outline: 0.125rem solid ", ";\n  }\n\n  :focus:not(:focus-visible) {\n    outline: none;\n  }\n"])), colors.font, typography.type.body, colors.primary, colors.primary, colors.shadow, colors.font, colors.primary, function (_a) {
    var disabled = _a.disabled;
    return disabled && " " + disabledStyles;
}, function (_a) {
    var errorMessage = _a.errorMessage;
    return !!errorMessage &&
        "border-color: " + colors.red + ";\n      box-shadow: none;\n  ";
}, colors.highlight);
export var StyledLabel = styled.label(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  display: block;\n  color: ", ";\n  margin-bottom: 0.3rem;\n  margin-left: 0.3rem;\n  font-size: 95%;\n  ", "\n"], ["\n  display: block;\n  color: ", ";\n  margin-bottom: 0.3rem;\n  margin-left: 0.3rem;\n  font-size: 95%;\n  ",
    "\n"])), colors.highlight, function (_a) {
    var disabled = _a.disabled;
    return disabled && disabledStyles + "; color: " + colors.disabled;
});
export var StyledTextareaError = styled.p(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  color: ", ";\n  font-size: 95%;\n  margin-top: 0.3rem;\n"], ["\n  color: ", ";\n  font-size: 95%;\n  margin-top: 0.3rem;\n"])), colors.red);
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
