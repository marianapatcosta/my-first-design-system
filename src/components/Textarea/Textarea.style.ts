import styled from 'styled-components';
import { baseStyles, disabledStyles, colors, typography } from '../../styles';

export const StyledTextareaWrapper = styled.div`
  ${baseStyles}
  margin: 1rem 0;
  position: relative;
  width: 100%;
`;

export const StyledTextarea = styled.textarea<{
  errorMessage?: string;
  disabled?: boolean;
}>`
  display: block;
  color: ${colors.font};
  font-family: ${typography.type.body};
  font-size: 95%;
  width: 100%;
  width: 100%;
  background-color: ${colors.primary};
  border: 0.125rem solid ${colors.primary};
  box-shadow: 0 0.063rem 0.25rem ${colors.shadow};
  box-sizing: border-box;
  padding: 0.3rem 0.625rem;
  position: relative;
  resize: none;

  :-webkit-autofill,
  :-webkit-autofill:hover,
  :-webkit-autofill:focus,
  :-webkit-autofill:active {
    -webkit-text-fill-color: ${colors.font};
    -webkit-box-shadow: 0 0 0 2rem ${colors.primary} inset !important;
  }

  ${({ disabled }) => disabled && ` ${disabledStyles}`}

  ${({ errorMessage }) =>
    !!errorMessage &&
    `border-color: ${colors.red};
  box-shadow: none;
  `}
`;

export const StyledLabel = styled.label<{ disabled?: boolean }>`
  display: block;
  color: ${colors.highlight};
  margin-bottom: 0.3rem;
  margin-left: 0.3rem;
  font-size: 95%;
  ${({ disabled }) =>
    disabled && `${disabledStyles}; color: ${colors.disabled}`}
`;

export const StyledTextareaError = styled.p`
  color: ${colors.red};
  font-weight: bold;
  font-size: 90%;
  margin-top: 0.3rem;
`;
