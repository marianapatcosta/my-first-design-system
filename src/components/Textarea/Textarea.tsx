import React, { ChangeEvent,  CSSProperties } from 'react';
import {
  StyledTextareaWrapper,
  StyledTextarea,
  StyledLabel,
  StyledTextareaError,
} from './Textarea.style';

export interface TextareaProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional css class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
  /**
   * Textarea label
   */
  label?: string;
  /**
   * Textarea name
   */
  name?: string;
  /**
   * Textarea value
   */
  value: string;
  /**
   * Textarea's displaying text when no option is selected
   */
  placeholder?: string;
  /**
   * Is Textarea disabled?
   */
  disabled?: boolean;
  /**
   * what is rows number allowed for the Textarea field?
   */
  rows?: number;
  /**
   * text to display in case of error
   */
  errorMessage?: string;
  /**
   * Change handler
   */
  onChange: (event: ChangeEvent<HTMLTextAreaElement>) => void;
  /**
   * Blur handler
   */
  onBlur?: (event: ChangeEvent<HTMLTextAreaElement>) => void;
}

export const Textarea: React.FC<TextareaProps> = ({
  id,
  label,
  disabled,
  errorMessage,
  ...props
}) => {
  return (
    <StyledTextareaWrapper>
      {label && (
        <StyledLabel htmlFor={id} disabled={disabled}>
          {label}
        </StyledLabel>
      )}
      <StyledTextarea
        data-testid='textarea-field'
        id={id}
        errorMessage={errorMessage}
        disabled={disabled}
        {...props}
      ></StyledTextarea>
      {!!errorMessage && (
        <StyledTextareaError>{errorMessage}</StyledTextareaError>
      )}
    </StyledTextareaWrapper>
  );
};

Textarea.defaultProps = {
  id: '',
  className: '',
  style: {},
  name: '',
  rows: 5,
  disabled: false,
  value: '',
  placeholder: '',
  errorMessage: '',
  onChange: () => null,
  onBlur: () => null,
};
