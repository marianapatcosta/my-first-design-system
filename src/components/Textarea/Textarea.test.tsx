import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Textarea } from './Textarea';

describe('Textarea', () => {
  const defaultProps = { label: 'textarea', value: 'Hello', onChange: () => null };
   test('renders without errors and matches snapshot', () => {
     const { container: { firstChild }  } = render(<Textarea  {...defaultProps}/>); 

     expect(firstChild).toMatchSnapshot();
   });

   test('should handle change events', () => {
    const onChangeMock = jest.fn();
    const { getByTestId } = render(
      <Textarea {...defaultProps} onChange={onChangeMock} />
    );
    const textarea = getByTestId('textarea-field') as HTMLInputElement;
    fireEvent.click(textarea, { target: { value: 'bye' } });
    expect(textarea.value).toBe('bye');
  });

  test('should handle blur events', () => {
    const onBlurMock = jest.fn();
    const { getByTestId } = render(
      <Textarea {...defaultProps} onBlur={onBlurMock} />
    );
    const textarea = getByTestId('textarea-field') as HTMLInputElement;
    fireEvent.blur(textarea);
    expect(onBlurMock).toBeCalled();
  });

  test('should return null on change if onChange prop is not passed', () => {
    const result =
    Textarea.defaultProps?.onChange && Textarea.defaultProps.onChange();
    expect(result).toBeNull();
  });

  test('should return null on blur if onChange prop is not passed', () => {
    const result = Textarea.defaultProps?.onBlur && Textarea.defaultProps.onBlur();
    expect(result).toBeNull();
  });

  describe('Disabled Textarea', () => {
    test('should not handle change events', () => {
      const onChangeMock = jest.fn();
      const {
        container: { firstChild },
      } = render(
        <Textarea {...defaultProps} disabled={true} onChange={onChangeMock} />
      );

      firstChild && fireEvent.click(firstChild);

      expect(onChangeMock).not.toBeCalled();
    });
  });

  describe('Textarea with error', () => {
    test('should render error message', () => {
      const { getByText } = render(
        <Textarea {...defaultProps} errorMessage='There is an error' />
      );

      const textarea = getByText(/There is an error/i);
      expect(textarea).toBeTruthy();
    });
  });
});