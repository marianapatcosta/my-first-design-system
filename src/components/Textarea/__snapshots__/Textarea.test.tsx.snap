// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea renders without errors and matches snapshot 1`] = `
<div
  class="sc-bdfBwQ jWcQjL"
>
  <label
    class="sc-dlfnbm iSaUXM"
    for=""
  >
    textarea
  </label>
  <textarea
    class="sc-gsTCUz iLUHcy"
    data-testid="textarea-field"
    id=""
    name=""
    placeholder=""
    rows="5"
  >
    Hello
  </textarea>
</div>
`;
